jetbrains-core/src/software/aws/toolkits/jetbrains/core/toolwindow/ToolkitToolWindow.kt

```diff
@@ -4,135 +4,86 @@
+import com.intellij.openapi.application.runInEdt
 import com.intellij.openapi.project.Project
-import com.intellij.openapi.util.Disposer
-import com.intellij.openapi.wm.ToolWindow
-import com.intellij.openapi.wm.ToolWindowAnchor
+import com.intellij.openapi.util.Key
 import com.intellij.openapi.wm.ToolWindowManager
 import com.intellij.ui.content.Content
-import com.intellij.ui.content.impl.ContentImpl
-import icons.AwsIcons
-import java.util.concurrent.ConcurrentHashMap
-import javax.swing.Icon
 import javax.swing.JComponent
 
 interface ToolkitToolWindow {
+    val project: Project
+    val toolWindowId: String
+
+    fun toolWindow() = ToolWindowManager.getInstance(project).getToolWindow(toolWindowId)
+        ?: throw IllegalStateException("Can't find tool window $toolWindowId")
+
     fun addTab(
         title: String,
         component: JComponent,
         activate: Boolean = false,
         id: String = title,
-        disposable: Disposable? = null,
-        refresh: () -> Unit = { }
-    ): ToolkitToolWindowTab
-
-    fun find(id: String): ToolkitToolWindowTab?
-
-    // prefix is prefix of the id. Assumes the window is using id composed of paths, like: "loggroup/logstream"
-    fun findPrefix(prefix: String): List<ToolkitToolWindowTab>
-}
-
-interface ToolkitToolWindowTab : Disposable {
-    val name: String
-    fun show(refresh: Boolean = false)
-}
-
-class ToolkitToolWindowManager(private val project: Project) : Disposable {
-    private val toolWindows = ConcurrentHashMap<ToolkitToolWindowType, ManagedToolkitToolWindow>()
-    internal fun getInstance(type: ToolkitToolWindowType) = toolWindows.computeIfAbsent(type) { ManagedToolkitToolWindow(type) }
-
-    inner class ManagedToolkitToolWindow(private val type: ToolkitToolWindowType) : ToolkitToolWindow {
-        private val tabs = mutableMapOf<String, ManagedToolkitToolWindowTab>()
-
-        override fun addTab(
-            title: String,
-            component: JComponent,
-            activate: Boolean,
-            id: String,
-            disposable: Disposable?,
-            refresh: () -> Unit
-        ): ToolkitToolWindowTab {
-            val content = ContentImpl(component, title, true)
-            val toolWindow = windowManager.getToolWindow(type.id)
-                ?: windowManager.registerToolWindow(type.id, true, type.anchor, this@ToolkitToolWindowManager, true).also {
-                    it.setIcon(type.icon)
-                    it.stripeTitle = type.title
-                }
-            Disposer.register(content, Disposable { closeWindowIfEmpty(toolWindow, type.id) })
-            disposable?.let { Disposer.register(content, it) }
-            toolWindow.contentManager.addContent(content)
-            return ManagedToolkitToolWindowTab(toolWindow, content, refresh).also {
-                tabs[id] = it
-                if (activate) {
-                    it.show()
-                }
+        disposable: Disposable? = null
```

Comment:
Should we force a `Disposable` instead of it being nullable?

Reply:
its an additional disposable

```
  @Override
  public void dispose() {
    if (myShouldDisposeContent && myComponent instanceof Disposable) {
      Disposer.dispose((Disposable)myComponent);
    }

    if (myDisposer != null) {
      Disposer.dispose(myDisposer);
      myDisposer = null;
    }

    myFocusRequest = null;
    clearUserData();
  }
```

Reply:
If content is already disposable, we dont need it.

Maybe should name it `additionalDisposable`?

Reply:
Renamed it, added KDoc